contracts/protocol/v1/traders/P1InverseOrders.sol

```diff
@@ -0,0 +1,495 @@
+     * @dev Emits the LogOrderFilled event.
+     *
+     * @param  sender  The address that called the trade() function on PerpetualV1.
+     * @param  maker   The maker of the order.
+     * @param  taker   The taker of the order.
+     * @param  price   The current oracle price of the underlying asset.
+     * @param  data    A struct of type TradeData.
+     * @return         The assets to be traded and traderFlags that indicate that a trade occurred.
+     */
+    function trade(
+        address sender,
+        address maker,
+        address taker,
+        uint256 price,
+        bytes calldata data,
+        bytes32 /* traderFlags */
+    )
+        external
+        returns (P1Types.TradeResult memory)
+    {
+        address perpetual = _PERPETUAL_V1_;
+
+        require(
+            msg.sender == perpetual,
+            "msg.sender must be PerpetualV1"
+        );
+
+        if (taker != sender) {
+            require(
+                P1Getters(perpetual).hasAccountPermissions(taker, sender),
+                "Sender does not have permissions for the taker"
+            );
+        }
+
+        TradeData memory tradeData = abi.decode(data, (TradeData));
+        bytes32 orderHash = _getOrderHash(tradeData.order);
+
+        // validations
+        _verifyOrderStateAndSignature(
+            tradeData,
+            orderHash
+        );
+        _verifyOrderRequest(
+            tradeData,
+            maker,
+            taker,
+            perpetual,
+            price
+        );
+
+        // set _FILLED_AMOUNT_
+        uint256 oldFilledAmount = _FILLED_AMOUNT_[orderHash];
+        uint256 newFilledAmount = oldFilledAmount.add(tradeData.fill.amount);
+        require(
+            newFilledAmount <= tradeData.order.amount,
+            "Cannot overfill order"
+        );
+        _FILLED_AMOUNT_[orderHash] = newFilledAmount;
+
+        // Inverse perpetual: The fill amount is denoted in quote currency (position).
+        emit LogOrderFilled(
+            orderHash,
+            tradeData.order.flags,
+            tradeData.order.triggerPrice,
+            tradeData.fill
+        );
+
+        // `isBuyOrder` is from the maker's perspective, and relative to the base currency.
+        bool isBuyOrder = _isBuy(tradeData.order);
+
+        // Inverse perpetual:
+        // - When isBuy is true, maker is buying base currency (margin) & selling quote (position).
+        // - The fill amount is denoted in quote currency (position).
+        // - Prices are denoted in quote currency per unit of base currency (position per margin).
+        // - The fee is paid (or received) in base currency (margin).
+        uint256 feeFactor = (isBuyOrder == tradeData.fill.isNegativeFee)
+            ? BaseMath.base().add(tradeData.fill.fee)
+            : BaseMath.base().sub(tradeData.fill.fee);
+        uint256 marginAmount = tradeData.fill.amount
+            .baseDiv(tradeData.fill.price)
+            .baseMul(feeFactor);
```

Comment:
nit: wonder if either of these would result in less rounding error

`tradeData.fill.amount.baseMul(feeFactor).baseDiv(tradeData.fill.price)`
`tradeData.fill.amount.mul(feeFactor).div(tradeData.fill.price)` (probably shouldn't overflow for most sane values)

Reply:
Yeah I think `tradeData.fill.amount.mul(feeFactor).div(tradeData.fill.price)` makes sense